source/geod24/concurrency.d

```diff
@@ -1665,12 +1756,58 @@ public class Channel (T)
 
     public T receive ()
     {
+        this.mutex.lock();
+
+        if (this.closed)
+        {
+            this.mutex.unlock();
+            assert(0, "Channel is closed.");
+        }
+
         T msg;
-        while (!tryReceive(&msg))
+
+        if (!this.sendq.empty)
         {
-            FiberScheduler.yield();
+            ChannelContext!T context = this.sendq.front;
+            this.sendq.removeFront();
+            msg = context.msg;
+            this.mutex.unlock();
+            context.notify();
+            return msg;
+        }
+
+        if (!this.queue.empty)
+        {
+            msg = this.queue.front;
+            this.queue.removeFront();
+            this.mcount--;
+            this.mutex.unlock();
+            return msg;
+        }
+
+        scope scheduler = thisScheduler;
+        if (Fiber.getThis() && scheduler)
+        {
+            ChannelContext!T new_context;
```

Comment:
These two `if` / `else` blocks again have similar code, so I suggest writing a helper function just like the one I suggested in the `send` routine.